charms/katib-controller/tests/integration/test_charm.py

```diff
@@ -0,0 +1,91 @@
+    "enasCPUTemplate": 'apiVersion: batch/v1\nkind: Job\nspec:\n  template:\n    spec:\n      containers:\n        - name: training-container\n          image: docker.io/kubeflowkatib/enas-cnn-cifar10-cpu:v0.15.0\n          command:\n            - python3\n            - -u\n            - RunTrial.py\n            - --num_epochs=1\n            - "--architecture=\\"${trialParameters.neuralNetworkArchitecture}\\""\n            - "--nn_config=\\"${trialParameters.neuralNetworkConfig}\\""\n      restartPolicy: Never',  # noqa: E501
+    "pytorchJobTemplate": 'apiVersion: kubeflow.org/v1\nkind: PyTorchJob\nspec:\n  pytorchReplicaSpecs:\n    Master:\n      replicas: 1\n      restartPolicy: OnFailure\n      template:\n        spec:\n          containers:\n            - name: pytorch\n              image: docker.io/kubeflowkatib/pytorch-mnist-cpu:v0.15.0\n              command:\n                - "python3"\n                - "/opt/pytorch-mnist/mnist.py"\n                - "--epochs=1"\n                - "--lr=${trialParameters.learningRate}"\n                - "--momentum=${trialParameters.momentum}"\n    Worker:\n      replicas: 2\n      restartPolicy: OnFailure\n      template:\n        spec:\n          containers:\n            - name: pytorch\n              image: docker.io/kubeflowkatib/pytorch-mnist-cpu:v0.15.0\n              command:\n                - "python3"\n                - "/opt/pytorch-mnist/mnist.py"\n                - "--epochs=1"\n                - "--lr=${trialParameters.learningRate}"\n                - "--momentum=${trialParameters.momentum}"',  # noqa: E501
+}
+
+
+@pytest.fixture(scope="session")
+def lightkube_client() -> lightkube.Client:
+    client = lightkube.Client(field_manager=CHARM_NAME)
+    return client
+
+
+class TestCharm:
+    @pytest.mark.abort_on_fail
+    async def test_build_and_deploy(self, ops_test: OpsTest):
+        """Build and deploy the charm.
+
+        Assert on the unit status.
+        """
+        charm_under_test = await ops_test.build_charm(".")
+        image_path = METADATA["resources"]["oci-image"]["upstream-source"]
+        resources = {"oci-image": image_path}
+
+        await ops_test.model.deploy(
+            charm_under_test, resources=resources, application_name=CHARM_NAME
+        )
+
+        await ops_test.model.wait_for_idle(
+            apps=[CHARM_NAME], status="active", raise_on_blocked=True, timeout=300
+        )
+
+    async def test_configmap_created(self, lightkube_client: lightkube.Client, ops_test: OpsTest):
+        """Test configmaps contents with default coonfig."""
+        katib_config_cm = lightkube_client.get(
+            ConfigMap, KATIB_CONFIG, namespace=ops_test.model_name
+        )
+        trial_template_cm = lightkube_client.get(
+            ConfigMap, TRIAL_TEMPLATE, namespace=ops_test.model_name
+        )
+
+        assert katib_config_cm.data == EXPECTED_KATIB_CONFIG
+        assert trial_template_cm.data == EXPECTED_TRIAL_TEMPLATE
```

Comment:
By testing the configmap contents, we're effectively mocking out Katib's interpretation of these configmaps and just asserting that they have the content we want.  Couldn't we do this in a unit test that asserts we `lightkube.apply()` with this content?

I wonder if at the integration tests we should instead try to **use** an image.  Create a Katib job that needs one of these images, and see if the generated Pods actually try to pull the custom image.  We don't need a real custom image somewhere that can be pulled, we could instead inspect the Pod yaml that Katib generates to make sure our custom image is being used.

Reply:
The unit tests will do the exact same thing as these integration ones. After discussion I will keep integration tests. 

The image usage thing from second part is (and originally was also) tested in the bundle tests which also deploy workloads. <3